biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java

```diff
@@ -209,6 +210,27 @@ public static ResolutionException augment(ResolveContext context, ResolutionExce
 		return augment(re.getUnresolvedRequirements(), context, re);
 	}
 
+	public static String format(Collection<Requirement> requirements) {
+		try (Formatter f = new Formatter()) {
+			for (Requirement req : requirements) {
+				boolean optional = isOptional(req);
+				f.format("%n %s[%-19s] %s%s", isOptional(req) ? " " : "*", req.getNamespace(), req,
+						isOptional(req) ? " ;resolution:=optional" : "");
+			}
+			return f.toString();
+		}
+	}
+
+	private static boolean isOptional(Requirement req) {
+		String resolution = req.getDirectives().get("resolution");
```

Comment:
Aren't there constant fields for these? Using constant fields makes it easier to find all the code using them.

https://github.com/bndtools/bnd/blob/master/biz.aQute.bndlib/src/aQute/bnd/osgi/Constants.java#L303